### ISSUE TITLE
Trainer forces the use of a specific collator

### ISSUE THREAD

**Issue by hteague-qti** [huggingface/trl#2481]
https://github.com/huggingface/trl/blob/c1e9ea8ecfdaf1486e2fc84bc55379b2338d5320/trl/trainer/gkd_trainer.py#L87

I don't understand why a trainer should force the use of a particular collator.   I means that I cannot just switch the trainer from SFT to GKD unless I have the same collator.  It's not clear what is special about the GKD trainer that means it can only work with one particular collator.

**Comment by kashif** [comment]
@hteague-qti so I wanted to get it working with this collator and then come back and make it more general after that.. so would you have a suggestion on what the next generalization could be? make it work for the SFT default collator?

**Comment by hteague-qti** [comment]
I was thinking it could be made completely independent of the collator.  First thing might be to warn users that even though they are providing a collator in the args, you are switching to a different one (for now).  

Seems to me that trainer should not care about the data preprocessing or the collator, just the output logits, etc.  Making it work with default collator in SFT would be fine.  This one is quite common for language: DataCollatorForCompletionOnlyLM

**Comment by hteague-qti** [comment]
btw, appreciate the response.

**Comment by YooSungHyun** [comment]
I have the same concerns, and if `padding_free` can be used, that would be ideal.  
However, I also find it somewhat challenging to figure out how to implement this.  
It seems that ChatML was chosen for implementation convenience, but it would be great if there are plans for an update.

@kashif

**Comment by albertvillanova** [comment]
Hey @hteague-qti, sorry for the very late follow-up — we're catching up on the backlog. 

This is resolved on current TRL: GKDTrainer now accepts a `data_collator` argument and only falls back to `DataCollatorForChatML` when you don't pass one (see https://github.com/huggingface/trl/blob/576a5f6bb536ca2fcfe934419d9f98f6f443eab8/trl/experimental/gkd/gkd_trainer.py#L132-L134), so you can drop in your own collator. Note the trainer has also moved to the experimental module — `from trl.experimental.gkd import GKDTrainer, GKDConfig`. 

Closing as fixed, but happy to reopen if a custom collator still doesn't behave for your case — the padding-free / packing angle @YooSungHyun raised is a separate, larger piece of work, so a focused issue for that would be welcome.
